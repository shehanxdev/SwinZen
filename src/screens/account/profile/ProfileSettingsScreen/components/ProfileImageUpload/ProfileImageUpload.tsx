import React, { useMemo, useState } from 'react';
import { ActivityIndicator, Alert, TouchableOpacity, View } from 'react-native';
import FastImage from 'react-native-fast-image';
import {
  CameraOptions,
  ImageLibraryOptions,
  ImagePickerResponse,
  launchCamera,
  launchImageLibrary,
} from 'react-native-image-picker';
import { useDispatch } from 'react-redux';

import { ProfileImageChangeCameraIcon, Text } from '@sz/components';
import { tw } from '@sz/config';
import { TextVariant } from '@sz/constants';
import { ToastService } from '@sz/services';
import { useSelector } from '@sz/stores';

//Note::below values are in PX
const PROFILE_IMAGE_DIMENTIONS = {
  height: 134,
  width: 134,
};
const CAMERA_ICON_DIMENTIONS = {
  height: 28,
  width: 28,
};
const PROFILE_IMAGE_COMMON_STYLES = tw`w-[${PROFILE_IMAGE_DIMENTIONS.width}px] h-[${PROFILE_IMAGE_DIMENTIONS.height}px] rounded-full bg-Primary-Sz650 items-center justify-center border-2 border-Neutral-Sz100`;
const CAMERA_ICON_POSITION = {
  top: PROFILE_IMAGE_DIMENTIONS.width / 8 - CAMERA_ICON_DIMENTIONS.height / 2,
  right: PROFILE_IMAGE_DIMENTIONS.width / 8 - CAMERA_ICON_DIMENTIONS.width / 2,
};
const CAMERA_OPTIONS: CameraOptions = {
  mediaType: 'photo',
  includeBase64: false,
  presentationStyle: 'fullScreen',
  saveToPhotos: false,
};
const GALLERY_OPTIONS: ImageLibraryOptions = {
  mediaType: 'photo',
  includeBase64: false,
  presentationStyle: 'fullScreen',
  selectionLimit: 1,
};

type ImagePickType = 'capture' | 'library';

export function ProfileImageUpload() {
  const [newProfileImageData, setNewProfileImageData] = useState<ImagePickerResponse>(null);

  const loading = useSelector(state => state.loading.effects.userStore.changeProfilePicture);

  const dispatch = useDispatch();

  const onButtonPress = async (type?: ImagePickType) => {
    let result: ImagePickerResponse;

    if (type === 'capture') {
      result = await launchCamera(CAMERA_OPTIONS);
    } else {
      result = await launchImageLibrary(GALLERY_OPTIONS, setNewProfileImageData);
    }

    if (result.didCancel) return;

    if (result.errorMessage) {
      ToastService.error({ message: 'Failed!', description: newProfileImageData?.errorMessage });
      return;
    }

    await dispatch.userStore.changeProfilePicture(result);
    setNewProfileImageData(result);
  };

  //TODO::fetch and display user data
  const renderProfileImage = useMemo(
    () =>
      newProfileImageData?.assets === undefined || loading ? (
        <View style={PROFILE_IMAGE_COMMON_STYLES}>
          {/* TODO::Create a helper function to get the first characters from full name when intergration with the APIs*/}
          {/* TODO::add a proper loading indicator */}
          {loading ? <ActivityIndicator /> : <Text variant={TextVariant.SubTitle1}>{'MW'}</Text>}
        </View>
      ) : (
        <FastImage
          style={PROFILE_IMAGE_COMMON_STYLES}
          source={{
            uri: newProfileImageData?.assets[0]?.uri,
          }}
        />
      ),
    [newProfileImageData, loading],
  );

  return (
    <View testID="ProfileImageComponentTestID" style={tw`m-auto`}>
      <TouchableOpacity
        onPress={() => {
          //TODO::provide a proper choose UI to the end user
          Alert.alert('Choose option', '', [
            {
              text: 'Camera',
              onPress: () => {
                onButtonPress('capture').catch(console.error);
              },
            },
            {
              text: 'Gallery',
              onPress: () => {
                onButtonPress().catch(console.error);
              },
            },
          ]);
        }}>
        {renderProfileImage}
        <View
          style={tw`w-[${CAMERA_ICON_DIMENTIONS.width}px] h-[${CAMERA_ICON_DIMENTIONS.height}px] rounded-full absolute right-[${CAMERA_ICON_POSITION.right}px] top-[${CAMERA_ICON_POSITION.top}px] justify-center items-center bg-Neutral-Sz100`}>
          <ProfileImageChangeCameraIcon />
        </View>
      </TouchableOpacity>
    </View>
  );
}
