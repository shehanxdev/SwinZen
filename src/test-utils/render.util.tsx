import { NavigationContext } from '@react-navigation/native';
import { render } from '@testing-library/react-native';
import React, { ReactNode } from 'react';
import Config from 'react-native-config';
import { Provider as PaperProvider } from 'react-native-paper';
import { Provider } from 'react-redux';

import { paperTheme } from '@sz/config';
import { ConfigService } from '@sz/services';
import { store } from '@sz/stores';

type RenderParams = Parameters<typeof render>;

export function renderWithProviders(ui: RenderParams[0], options?: RenderParams[1]) {
  ConfigService.setCallback(key => Config[key]);

  const BaseProviders = ({ children }: { children: ReactNode }) => (
    <Provider store={store}>
      <NavigationContext.Provider
        //@ts-ignore
        value={{
          isFocused: () => true,
          addListener: jest.fn(() => jest.fn()),
        }}>
        <PaperProvider theme={paperTheme}>{children}</PaperProvider>
      </NavigationContext.Provider>
    </Provider>
  );

  return render(ui, { wrapper: BaseProviders, ...options });
}
