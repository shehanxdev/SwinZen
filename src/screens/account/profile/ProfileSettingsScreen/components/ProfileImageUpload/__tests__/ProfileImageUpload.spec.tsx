import * as React from 'react';

import { renderWithProviders } from '@sz/test-utils';

import { ProfileImageUpload } from '../ProfileImageUpload';

describe('Profile Image Upload component', () => {
  const testID = 'ProfileImageComponentTestID';
  const getRenderedComponent = () => renderWithProviders(<ProfileImageUpload />);

  it(`should render component correctly`, () => {
    const renderer = getRenderedComponent();
    const renderTree = renderer.toJSON();
    expect(renderTree).toMatchSnapshot();
  });

  it('should find the component via testID', () => {
    const { getByTestId } = getRenderedComponent();
    const foundScreen = getByTestId(testID);
    expect(foundScreen).toBeTruthy();
  });
});
