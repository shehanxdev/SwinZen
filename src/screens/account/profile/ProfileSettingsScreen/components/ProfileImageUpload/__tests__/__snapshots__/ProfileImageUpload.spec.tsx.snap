// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Image Upload component should render component correctly 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "marginBottom": "auto",
        "marginLeft": "auto",
        "marginRight": "auto",
        "marginTop": "auto",
      }
    }
    testID="ProfileImageComponentTestID"
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#1A5C23",
            "borderColor": "#F8F9FA",
            "borderRadius": 9999,
            "borderWidth": 2,
            "height": 134,
            "justifyContent": "center",
            "width": 134,
            "zIndex": 0,
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "#F8F9FA",
                "lineHeight": 40,
              },
              {
                "fontFamily": "SourceSansPro-Regular",
                "fontSize": 28,
                "fontWeight": "600",
                "textAlign": "center",
              },
            ]
          }
        >
          SZ
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#F8F9FA",
            "borderRadius": 9999,
            "height": 28,
            "justifyContent": "center",
            "position": "absolute",
            "right": 2.75,
            "top": 2.75,
            "width": 28,
            "zIndex": 20,
          }
        }
      >
        <SvgMock
          fill="#3C563E"
          height={10}
          width={14}
        />
      </View>
    </View>
    <View
      testID="ImageUplaodModalTestID"
    >
      <Modal
        animationType="none"
        deviceHeight={null}
        deviceWidth={null}
        hardwareAccelerated={false}
        hideModalContentWhileAnimating={false}
        onBackdropPress={[Function]}
        onModalHide={[Function]}
        onModalWillHide={[Function]}
        onModalWillShow={[Function]}
        onRequestClose={[Function]}
        panResponderThreshold={4}
        scrollHorizontal={false}
        scrollOffset={0}
        scrollOffsetMax={0}
        scrollTo={null}
        statusBarTranslucent={false}
        supportedOrientations={
          [
            "portrait",
            "landscape",
          ]
        }
        swipeThreshold={100}
        transparent={true}
        visible={false}
      />
    </View>
  </View>
</View>
`;
